text/0035-opentelemetry-protocol.md

```diff
@@ -82,23 +82,76 @@ If the client is shutting down (e.g. when the containing process wants to exit)
+When using gRPC transport success response is returned via `ExportResponse` message.
 
-- `FailedRetryable` - processing of telemetry data failed. The client SHOULD record the error and may retry exporting the same data immediately. This can happen when the server is temporarily unable to process the data.
+When an error is returned by the server it falls into 2 broad categories: retryable and not-retryable.
+
+- Retryable errors indicate that processing of telemetry data failed and the client SHOULD record the error and may retry exporting the same data. This can happen when the server is temporarily unable to process the data.
+
+- Not-retryable errors indicate that processing of telemetry data failed and the client MUST NOT retry sending the same telemetry data. The telemetry data MUST be dropped. This for example can happen when the request contains bad data and cannot be deserialized or otherwise processed by the server. The client SHOULD maintain a counter of such dropped data.
+
+When using gRPC transport the server SHOULD indicate retryable errors using code [Unavailable](https://godoc.org/google.golang.org/grpc/codes) and MAY supply additional [details via status](https://godoc.org/google.golang.org/grpc/status#Status.WithDetails) using [RetryInfo](https://github.com/googleapis/googleapis/blob/6a8c7914d1b79bd832b5157a09a9332e8cbd16d4/google/rpc/error_details.proto#L40) containing 0 value of RetryDelay. Here is a sample Go code to illustrate:
+
+```go
+  // Do this on server side.
+  st, err := status.New(codes.Unavailable, "Server is unavailable").
+    WithDetails(&errdetails.RetryInfo{RetryDelay: &duration.Duration{Seconds: 0}})
+  if err != nil {
+    log.Fatal(err)
+  }
+
+  return st.Err()
+```
+
+To indicate not-retryable errors the server is recommended to use code [InvalidArgument](https://godoc.org/google.golang.org/grpc/codes) and MAY supply additional [details via status](https://godoc.org/google.golang.org/grpc/status#Status.WithDetails) using [BadRequest](https://github.com/googleapis/googleapis/blob/6a8c7914d1b79bd832b5157a09a9332e8cbd16d4/google/rpc/error_details.proto#L119). Other gRPC status code may be used if it is more appropriate. Here is a sample Go code to illustrate:
+
+```go
+  // Do this on server side.
+  st, err := status.New(codes.InvalidArgument, "Invalid Argument").
+    WithDetails(&errdetails.BadRequest{})
+  if err != nil {
+    log.Fatal(err)
+  }
+
+  return st.Err()
+```
+
+The server MAY use other gRPC codes to indicate retryable and not-retryable errors if those other gRPC codes are more appropriate for a particular erroneous situation. The client SHOULD interpret gRPC status codes as retryable or not-retryable according to the following table:
+
+|gRPC Code|Retryable?|
+|---------|----------|
+|CANCELLED|Yes|
+|UNKNOWN|No|
+|INVALID_ARGUMENT|No|
+|DEADLINE_EXCEEDED|No|
```

Comment:
This should be retryable because it means you hit a high loaded server?

Reply:
I am torn on this one. Here is what the code says:
>The deadline expired before the operation could complete. For operations that change the state of the system, this error may be returned even if the operation has completed successfully. For example, a successful response from a server could have been delayed long

So it says _"this error may be returned even if the operation has completed successfully"_. This means that if we retry we may create duplicates of data and potentially this can repeat for many times if the reason for DEADLINE_EXCEEDED is because of the specific data that we send (e.g. data is composed such that it takes a long time to process). We may end up in infinite loop retrying the same data over and over again while the data is actually successfully processed.

Thoughts?

Reply:
There are multiple scenarios in which deadline exceeded can be returned:
* Request did get to the server, request got processed, and got canceled before sending the response.
* Request did get to the server, and got canceled half way through processing the request.
* Request did get to the server, and got canceled before processing the request (waiting in the processing queue).
* Request did not get to the server (waiting in client sending queue) and got canceled.

I think in at least half of these cases we want to retry with a backoff.

Reply:
Yeah, as I mentioned I am torn on this one. If you feel strong that it should be retryable then we can mark it as retryable, but we need it to be marked one way or another. We can't have both unless we require extra details with the status and I think we shouldn't since it still may be returned without status details and we still need to make a decision.
I am fine with settling on making DEADLINE_EXCEEDED retryable. In practice I don't think this is going to be a big deal and if it is we can rethink it.

Reply:
Marking this retryable seems fine to me. Maybe it's still treated as a special case. There's nothing about DEADLINE_EXCEEDED that says you shouldn't try again, it suggests it may fail again though. The server should do it's best to not to let this response through by some sort of early failure, right?

Reply:
I changed DEADLINE_EXCEEDED to be retryable.